File: Microsoft/172.factorial-trailing-zeroes.py
"""
Approach: Count factors of 5 in n!.
Data structure: a running integer count is sufficient.
Interview logic: each trailing zero comes from a factor pair 2*5, and factorials contain more factors of 2 than 5. Numbers like 25 contribute multiple factors of 5, handled by repeatedly dividing n by 5.
Complexity: O(log_5 n) time, O(1) space.
Tests and edge cases: n < 5 returns 0; powers of 5 add extra counts; large n remains fast.
"""
from __future__ import annotations

# @lc code=start
class Solution:
    def trailingZeroes(self, n: int) -> int:
        count = 0
        while n:
            n //= 5
            count += n
        return count
# @lc code=end

File: Google/172.factorial-trailing-zeroes.java
/**
 * Algorithm:
 * Trailing zeroes come from factors of 10, and factors of 5 are scarcer than
 * factors of 2. Count multiples of 5, 25, 125, and so on.
 *
 * Complexity:
 * Time O(log_5 n), space O(1).
 */
class Solution {
    public int trailingZeroes(int n) {
        int count = 0;
        while (n != 0) {
            n /= 5;
            count += n;
        }
        return count;
    }
}
